### PR TITLE
Switch to futures + deltalake

### PR DESCRIPTION
**Pull request by mrocklin**
cc @jrbourbeau

### CODE REVIEW

File: preprocess.py [hunk 1]
```diff
@@ -145,16 +173,33 @@ def list_files(start, stop):
...
+            })
         with cluster.get_client() as client:
+            client.restart()
             futures = client.map(process_file, filenames)
-            for future, result in as_completed(futures, with_results=True):
-                print(".", end="")
-                future.cancel()
-                write_delta(result)
+            futures = client.map(write_delta, futures, retries=10)
+            wait(futures)
```
Comment (mrocklin) [inline preprocess.py:205]: I'm curious, any collisions arise here?

Comment (jrbourbeau) [reply preprocess.py:205]: There is collision here, but `retries=10` makes this work smooth consistently at the `n_workers=100` scale. That seemed good enough for me at the current stage we're at

File: preprocess.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    )
+
+    with Cluster() as cluster:
+        if not LOCAL:
+            cluster.send_private_envs({
+                "AWS_ACCESS_KEY_ID": os.environ["AWS_ACCESS_KEY_ID"],
+                "AWS_SECRET_ACCESS_KEY": os.environ["AWS_SECRET_ACCESS_KEY"],
+            })
+        with cluster.get_client() as client:
+            client.restart()
```
Comment (jrbourbeau) [inline preprocess.py:202]: Whoops, this is from debugging locally

File: preprocess.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    return filenames
+
+
+def compact(table):
+    t = deltalake.DeltaTable(
+        OUTDIR / table,
+        storage_options=STORAGE_OPTIONS,
+    )
+    t.optimize.compact()
+    t.vacuum(retention_hours=0, enforce_retention_duration=False, dry_run=False)
```
Comment (jrbourbeau) [inline preprocess.py:185]: Noting that this vacuum call doesn't seem to be working. When I `aws s3 ls` the data directory I still see all the original small files. Not sure what's going on there at the moment.

File: preprocess.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    Cluster = LocalCluster
+else:
+    OUTDIR = Path(f"s3://openscapes-scratch/etl-github")
+    STORAGE_OPTIONS = {"AWS_REGION": "us-west-2", "AWS_S3_ALLOW_UNSAFE_RENAME": "true"}
+    Cluster = functools.partial(
+        coiled.Cluster,
+        name="etl-github",
+        n_workers=50,
+        region="us-west-2",
+        shutdown_on_close=False,
```
Comment (jrbourbeau) [inline preprocess.py:31]: Note I'm keeping the cluster up for rapid iteration on the script. Not a long term thing.

File: preprocess.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        start=datetime.date(2024, 1, 1),
+        stop=datetime.date(2024, 1, 10),
+    )
+
+    with Cluster() as cluster:
+        if not LOCAL:
+            cluster.send_private_envs({
+                "AWS_ACCESS_KEY_ID": os.environ["AWS_ACCESS_KEY_ID"],
+                "AWS_SECRET_ACCESS_KEY": os.environ["AWS_SECRET_ACCESS_KEY"],
+            })
```
Comment (jrbourbeau) [inline preprocess.py:200]: It's unfortunate this is needed, but delta needs these environment variables to be set

**Pull request merged**